### PR TITLE
[Merged by Bors] - feat: Add `Zlattice.basis`

### PR DESCRIPTION
**Pull request by xroblot**
For a  `ℤ`-lattice `L` of `E`, defined as a discrete subgroup that spans the whole space `E` over `K`, proves that any  `ℤ`-basis of `L` is also a `K`-basis of `E`. This provides the link between the two points of view on lattices in this file. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by riccardobrasca** [comment]
Can you add that any `ℤ`-basis of `L` gives a `K`-basis of `E`? Both the results of this PR seem to generalize to this setting (we want at least the first one to easily get such a basis):

**Comment by xroblot** [comment]
> Can you add that any `ℤ`-basis of `L` gives a `K`-basis of `E`? Both the results of this PR seem to generalize to this setting (we want at least the first one to easily get such a basis):

Yes, but I think I can only do it for `K = ℝ` since otherwise Lean does not know that `L` admits a basis. (The instances are only for the real case.)

**Comment by riccardobrasca** [comment]
> > Can you add that any `ℤ`-basis of `L` gives a `K`-basis of `E`? Both the results of this PR seem to generalize to this setting (we want at least the first one to easily get such a basis):
> 
> Yes, but I think I can only do it for `K = ℝ` since otherwise Lean does not know that `L` admits a basis. (The instances are only for the real case.)

I mean defining a function that takes a basis of `L` and gives a basis of `E` such that the span is `L` or something like that.

**Comment by xroblot** [comment]
> > > Can you add that any `ℤ`-basis of `L` gives a `K`-basis of `E`? Both the results of this PR seem to generalize to this setting (we want at least the first one to easily get such a basis):
> > 
> > 
> > Yes, but I think I can only do it for `K = ℝ` since otherwise Lean does not know that `L` admits a basis. (The instances are only for the real case.)
> 
> I mean defining a function that takes a basis of `L` and gives a basis of `E` such that the span is `L` or something like that.

Oh, my mistake. I thought it was necessary to have a free module to talk about bases, but that it is not the case...

**Comment by riccardobrasca** [comment]
> > > > Can you add that any `ℤ`-basis of `L` gives a `K`-basis of `E`? Both the results of this PR seem to generalize to this setting (we want at least the first one to easily get such a basis):
> > > 
> > > 
> > > Yes, but I think I can only do it for `K = ℝ` since otherwise Lean does not know that `L` admits a basis. (The instances are only for the real case.)
> > 
> > 
> > I mean defining a function that takes a basis of `L` and gives a basis of `E` such that the span is `L` or something like that.
> 
> Oh, my mistake. I thought it was necessary to have a free module to talk about bases, but that it is not the case...

I perfectly understand your sentence, but I find it hilarious :D

File: Mathlib/Algebra/Module/Zlattice.lean [hunk 1]
```diff
@@ -552,4 +552,43 @@ theorem Zlattice.rank [hs : IsZlattice K L] : finrank ℤ L = finrank K E := by
     rw [← topEquiv.finrank_eq, ← h_spanE]
     convert finrank_span_le_card (R := K) (Set.range b)
 
+open Module
+
+/-- Any `ℤ`-basis of `L` is also a `K`-basis of `E`. -/
+def Basis.ofZlatticeBasis {ι : Type*} [hs : IsZlattice K L] (b : Basis ι ℤ L) :
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Module/Zlattice.lean:558]: You can declare `ι` and `b` just once.

**Pull request closed without merging**